### PR TITLE
#163388144 Add User Notifications

### PR DESCRIPTION
**Pull request by Arusey**
**What does this PR do?**
-Sends email notifications to followers whenever an article is created
-Sends In-app notifications to followers whenever an article is created
**Description of Task to be completed?**
-Users can subscribe to both in-app and email notifications
-Users can unsubscribe to both in-app and email notifications
-Followers can receive both in-app and email notifications whenever an article is created
**How should this be manually tested?**
* checkout the branch `ft-create-user-notifications-163388144`
* run migrations `python manage.py migrate`
* after running the server signup two users
* user 1 follow user 2
* user 1 unsubscribe from notifications `api/notifications/unsubscribe`
* User 2 create an article
* User 1 check notifications `api/notifications`
* User 1 should have no notifications
* user 1 subscribe for notifications `api/notifications/subscribe`
* User 2 Create an article
* User 1 should receive notifications
**What are the relevant pivotal tracker stories?**
[#163388144](https://www.pivotaltracker.com/n/projects/2240061) 
**Screenshots to support the PR**
1. Email notification
image
2.In-app Notification
image
3. Subscribe to notifications
image
4. Unsubscribe to notifications
image
image

### CODE REVIEW

File: authors/apps/authentication/serializers.py [hunk 1]
```diff
@@ -281,3 +281,10 @@ class Meta:
         fields = ('username', 'bio', 'image',
                   'created_at', 'updated_at')
         read_only_fields = ('username', 'created_at', 'updated_at',)
+
+
+class SubscriptionSerializer(serializers.Serializer):
+
+    def validate(self, validated_data):
+        print(validated_data)
```
Comment (dev-jey) [inline authors/apps/authentication/serializers.py]: remove print statement

File: authors/apps/notify/serializers.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+
+class NotificationSerializer(serializers.ModelSerializer):
+    '''
+    notification serializer
+    '''
+    actor = UserSerializer('actor_object_id')
+    action_object = ArticleSerializer('action_object_object_id')
+    recipient = UserSerializer(User, read_only=True)
+    # unread = serializers.BooleanField(read_only=True)
```
Comment (dev-jey) [inline authors/apps/notify/serializers.py]: remove commented code

Comment (Arusey) [reply authors/apps/notify/serializers.py]: done

File: authors/apps/authentication/serializers.py (hunk 1 above)
Comment (Arusey) [reply authors/apps/authentication/serializers.py]: done

File: authors/apps/authentication/views.py [hunk 3]
```diff
@@ -446,3 +449,47 @@ def get(self, request):
...
+
+
+class UnsubscribeAPIView(APIView):
+    "Allows user to unsubscribe to notifications"
+
+    permission_classes = (AllowAny,)
+
+    serializer_class = SubscriptionSerializer
+
+    def get(self, request, uuid):
```
Comment (asirvex) [inline authors/apps/authentication/views.py]: Using GET for unsubscribing is a bit confusing. I suggest you use put

File: authors/apps/articles/models.py [hunk 4]
```diff
@@ -57,3 +66,42 @@ class Rating(models.Model):
...
+    """Create a Signal that sends email to all users that follow the author.
+     Arguments:
+        sender {[type]} -- [Instance of ]
+        created {[type]} -- [If the article is posted.]
+    """
+
+    if instance and created:
+        users_followers = instance.author.followers.all()
+
+        link = f'{os.getenv("HEROKU_BACKEND_URL")}/api/v1/articles/' + \
```
Comment (sirghiny) [inline authors/apps/articles/models.py]: You can use triple quotes on f-strings to use them over multiple lines. Check: https://www.reddit.com/r/Python/comments/7kajpm/how_to_do_multiline_f_string_without_the/

File: authors/apps/authentication/views.py (hunk 3 above)
Comment (Arusey) [reply authors/apps/authentication/views.py]: done

File: authors/apps/articles/models.py (hunk 4 above)
Comment (Arusey) [reply authors/apps/articles/models.py]: Done pushing changes right now

**Pull request merged**